Sys Runtime Configuration Registry

LasseRosenow opened the pull request:
# RIOT Runtime Configuration Registry

This is a continuation of the effort previously done in #10622, but its architecture has dramatically changed since then.

## Abstract

This PR implements a system level runtime configuration system for RIOT.

A runtime configuration system is in charge of providing a mechanism to set and get the values of `Configuration Parameters` that are used during the execution of the firmware, as well as a way to persist these values. Runtime configurations are deployment-specific and can be changed on a per node basis.
Appropriate management tools could also enable the configuration of nodes.

Examples of runtime configurations are:

- Transmission duty cycles
- Sensor thresholds
- Security credentials
- System state variables
- RGB-LED colors

These parameters might have constraints, like a specific order to be applied (due to interdependencies) or value boundaries.

The main advantages of having such a system are:

- Easy to apply per-node configuration during deployment
- No need to implement a special mechanism for per-node configurations during firmware updates (only in the case of migration), as the parameters persist.
- Common interface for modules to expose their runtime Configuration Parameters and handle them
- Common interface for storing Configuration Parameters in non-volatile storage devices

## Modules

The RIOT Registry consists of various modules that are not all needed for minimal operation:

**`registry`**\
Core functionality such as set, get, export and apply operations to retrieve and set configurations.

**`registry_storage`**\
Adds capability to safe parameters to non-volatile storage devices.

**`registry_int_path`**\
Allows accessing configuration parameters via a path of integers.

**`registry_string_path`**\
Allows accessing configuration parameters via a path of strings.

**`registry_cli`**\
Allows accessing configuration parameters via a shell command.

## Design

### Architecture

The proposed architecture, as shown below, is formed by one or more Applications or Configuration Managers and the RIOT Registry.
The RIOT Registry acts as a common interface to access Runtime Configurations and store them in non-volatile storage.
All runtime configurations can be accessed either from the RIOT application using the provided RIOT Registry interfaces or through the interfaces exposed by the Configuration Managers.
A RIOT Application may interact with a Configuration Manager in order to modify access control rules or enable different exposed interfaces.

#### Path Based Configuration Managers (Needs `int_path or string_path` extension)

These Configuration Managers are a simple representation of the default configuration structure of the RIOT Registry.
They use either the `int_path` or the `string_path` Registry extension module to expose the parameters using a path.

#### Custom Schema Based Configuration Managers

These Configuration Managers have their own configuration structure (custom predefined object models etc.) and can not automatically be mapped to / from the RIOT Registry itself.
To make them work, a custom mapping module needs to be implemented, which maps each Configuration Parameter from the registry to the correct format of the Configuration Manager.

![design_architecture](https://github.com/RIOT-OS/RIOT/assets/10547444/26830a26-882a-422c-ba20-abfa4e07ce8c)

### Namespaces and Storages

The RIOT Registry interacts with RIOT modules via `Configuration Schemas`, and with non-volatile storages via `Storages`.
This way the functionality of the RIOT Registry is independent of the functionality of a `module` or `storage` implementation.
It is possible to get or set the values of `Configuration Parameters`.
It is also possible to transactionally apply configurations or export their values to a buffer or print them.
To persist Configuration Values, it is possible to store them in non-volatile storages.

Any mechanism of security (`access control`, `encryption` of configurations) is `not` directly in the scope of the Registry but in the Configuration Managers and the specific implementations of the `Configuration Schemas` and `Storages`.

The graphic below shows an example of two `Configuration Namespaces` (SYS and APP).
The `APP` namespace contains a application specific `My app` Configuration Schema and the `SYS` namespace specifies a `WLAN` and a `LED Strip` Configuration Schema.
The application `My app` uses the custom `My app` Configuration Schema to expose custom Configuration Parameters to the RIOT Registry and the drivers `WS2812`, `SK6812` and `UCS1903` contain instances of
the `LED Strip` Configuration Schema to expose common LED Strip Configuration Parameters.

Also, there are two Storages available: `MTD` and `VFS`.
The `MTD` Storage internally uses the RIOT `MTD` driver and the `VFS` Storage internally uses the RIOT `VFS` module.

![design_namespaces_and_storages](https://github.com/user-attachments/assets/49b066bc-5419-454a-9797-c0972b3bc83b)

### Components

The RIOT Registry is split into multiple components as can be seen in the graphic below:

#### Registry Core

This component holds the most basic functionality of the RIOT Registry.
It allows to `set` and `get` Configuration Values, transactionally `apply` them to make the changes come into effect and `export` all Configuration Parameters that exist in a given `Configuration Namespace`, `Configuration Schema` or `Configuration Group`.

Furthermore it is possible to `add` `Configuration Namespaces` or `Configuration Schema Instances`.

#### Registry Namespace

The `Configuration Namespaces` such as `SYS` or `APP` and their respective `Configuration Schemas` are not part of the Registry itself.
It is possible to `add` custom `Configuration Namespaces` depending on the given needs.

#### Storage

The `Storage` component provides an interface to `load` Configuration Values from a persistent `Storage` implementation or to `save` the current Registry configuration to it.

#### Registry Storage [Extension]

The implementations of a Storage such as `VFS` or `MTD` are not part of the Registry itself and can be switched out with implementations that are most suitable to the given needs.

#### Integer Path [Extension]

The `int_path` component provides helper functions that convert a path of up to 4 integer values to the respective pointer of a `Configuration Namespace`, `Configuration Schema`, `Configuration Schema Instance`, `Configuration Group` or `Configuration Parameter` and the other way around.

The structure of an integer configuration path is the following:\
`namespace_id` / `schema_id` / `instance_id` / `group_id | parameter_id`

**For example:**

| Path | Result |
|---|---|
| 0 | Namespace Object |
| 0 / 1 | Namespace and Schema Object |
| 0 / 1 / 0 | Namespace, Schema and Instance Object |
| 0 / 1 / 0 / 3 | Namespace, Schema, Instance and Group or Parameter Object |

#### String Path [Extension]

The `string_path` component provies helper functions that convert a string path to the respective pointer of a `Configuration Namespace`, `Configuration Schema`, `Configuration Schema Instance`, `Configuration Group` or `Configuration Parameter` and the other way around.

The structure of a string configuration path is the following:\
`namespace_name` / `schema_name` / `instance_id` (/ `group_name`)* / `parameter_name`.

The amount of path items is flexible, so the path could only consist of the `namespace_name` or only the `namespace_name` and the `schema_name` and so on.

**For example:**\

| Path | Result |
|---|---|
| sys | Namespace Object |
| sys / temperature_pressure_humidity | Namespace and Schema Object |
| sys / temperature_pressure_humidity / 0 | Namespace, Schema and Instance Object |
| sys / temperature_pressure_humidity / 0 / calibration | Namespace, Schema, Instance and Group Object |
| sys / temperature_pressure_humidity / 0 / calibration / humidity | Namespace, Schema, Instance, Group and Parameter Object |
| sys / temperature_pressure_humidity / 0 / last_reading_timestamp | Namespace, Schema, Instance and Parameter Object |

![design_registry](https://github.com/user-attachments/assets/473cef24-39b8-4539-b754-53146a497521)

## API

The graphic below shows the API of the RIOT Registry.
The top shows the Core API to manage Configuration Parameters.
On the right-hand side are functions to `set` and `get` Configuration Parameters, transactionally `apply` them and `export` them to a buffer or terminal.
On the left-hand side are setup functions to `add` `Configuration Namespaces` and `Configuration Schema Instances` to the Registry.

The bottom shows the storage API to manage the persistance of Configuration Parameters.
The left-hand side shows functions to `load”` and `save` Configuration Parameters to and from the persistent Storage.
The right-hand side shows functions to `add` `Storage Sources` (for reading) and to `set` a `Storage Destination` (for writing).
The Registry can have multiple `Storage Sources`, but always only one `Storage Destination`.
This allows to migrate from an old `Storage` to a new one.

The functionality of these functions is explained in the following paragraphs.

![api_structure](https://github.com/user-attachments/assets/588cb389-250a-4797-a78e-a76c4efe18fb)

### Core API

#### Get

A Configuration Value can be retrieved using the `registry_get` function.
The function takes the `registry_node_t` and a `registry_value_t` pointer (to return the value) as its arguments.

```c
int registry_get(
    const registry_node_t *node,
    registry_value_t *value
);
```

#### Set

A Configuration Value can be set using the `registry_set` function.
The function takes the `registry_node_t`, a `void*` buffer and the buffer size as its arguments.

The buffer must contain the value in its correct c-type.
If the Registry expects a `u8`, but a `u16` is provided, the operation will fail.
Furthermore the registry can specify constraints like `minimum` and `maximum` values and an array of `allowed` or `forbidden` values.
If these constraints are not fulfilled, then the operation will fail as well.

```c
int registry_set(
    const registry_node_t *node,
    const void *buf,
    const size_t buf_len,
);
```

#### Apply

Once the value(s) of one or multiple `Configuration Parameter(s)` are changed by the `registry_set` function, they still need to be applied, so that the new values are taken into effect.

Configuration Parameter(s) can be applied using the `registry_apply` function.
In this case the Registry provides multiple apply functions to allow applying in varying degrees.
The provided functions are `registry_apply`, this function applies every `Configuration Parameter` currently available in the Registry within the specified scope, specified by the `registry_node_t` argument.

When a whole `Schema Instance`, or a single `Configuration Parameter` is applied, it will be passed on to the `apply_cb` handler of the `Configuration Schema Instance`, provided by the module that needs runtime configuration.
This way the module gets notified, when the `Configuration Parameter` has been applied and can apply the changes accordingly.

```c
int registry_apply(const registry_node_t *node);
```

#### Export

Some times it is convenient to have a way to see what `Configuration Namespaces`, `Configuration Schemas`, `Configuration Schema Instances`, `Configuration Groups` or `Configuration Parameters` are available within our current RIOT Registry deployment.
To get this information there is the `registry_export` function.

This function exports every `Configuration Object` currently available in the Registry within the scope of the provided `registry_node_t` argument.

When a node in the schema is exported, it will be passed on to the `export_cb` handler provided as an argument of each `registry_export*` function.

```c
int registry_export(
    const registry_node_t *node,
    const registry_export_cb_t export_cb,
    const uint8_t tree_traversal_depth,
    const void *context
);
```

#### Add Namespaces to the Registry

To be able to use `Configuration Schemas` and their `Parameters` etc. it is necessary to add a `Configuration Namespace` that holds the required `Configuration Schemas` to the Registry.

This is possible using the `REGISTRY_ADD_NAMESPACE` macro, providing the name of the `Configuration Namespace` and a pointer to a `registry_namespace_t` object as arguments.

```c
#define REGISTRY_ADD_NAMESPACE(_name, _namespace)
```

#### Add Configuration Schema Instances to the Registry

To implement runtime configuration functionality into a module, it is necessary to add a `Configuration Schema Instance` of the needed `Configuration Schema` to the Registry.

This is possible using the `registry_add_schema_instance` function, providing the `Configuration Schema` and the `Configuration Schema Instance` as arguments.

```c
int registry_add_schema_instance(
    const registry_schema_t *schema,
    const registry_instance_t *instance,
);
```

### Storage API (Extension)

#### Load from Storage

It is often needed to load Configuration Parameters from a non-volatile `Storage` device.
For example when a device restarts after a shutdown.

This is possible using the `registry_storage_load` function.
This function takes a `registry_storage_instance_t` as its arguments.
The `the registry_storage_instance_t` contains data such as the mount point.
Internally the `Storage Instance` searches its persistent storage device for `Configuration Values`.
If a `Configuration Value` is found, the `Storage Source` calls the `registry_set` function and provides the necessary arguments.

```c
int registry_storage_load(
    const registry_storage_instance_t *storage_instance,
);
```

#### Save to Storage

To save `Configuration Values` to a non-volatile `Storage` device, the Registry provides the `registry_storage_save` function.
This function saves all available `Configuration Parameters` within the scope specified by the `registry_storage_node_t` argument.

Internally these functions call the `save` handler of the provided `Storage Instance` for each `Configuration Parameter` that has to be saved to storage.
The `save` handler of the `Storage`, of the provided `Storage Instance` takes a `Storage Instance` providing data such as the mount point, a `registry_node_t`, which describes the location inside the configuration tree and a `Configuration Value` as its arguments.
These values provide the needed information to load them back into the registry.

The way how the `Storage` stores these values internally is not specified.

```c
int registry_storage_save(
    const registry_storage_instance_t *storage_instance,
    const registry_node_t *node,
);
```

#### Set Storage Instances

To be able to expose storage instances to configuration managers such as the RIOT CLI, we have a function called `registry_storage_set_instances`, which lets you provide an array of storage instances.

```c
registry_error_t registry_storage_set_instances(
    const registry_storage_instance_t **storage_instances,
);
```

#### Get Storage Destination

To be able to retrieve a list of existing storage instances, we have a function called `registry_storage_get_instances`, which takes a pointer to an array of `storage_instance_t` structs as an output parameter.

```c
registry_error_t registry_storage_get_instances(
    const registry_storage_instance_t ***storage_instances,
);
```

## Comparison to Apache Mynewt Config

While originally our work on the RIOT Registry was heavily inspired by Apache Mynewt Config, it has since evolved to provide features such as `Type Safety`, make `String Paths` optional, introduce `Integer Paths` and provide a more modular `Pointer-Based API`.

The table below shows the difference between Apache Mynewt Config and the proposed RIOT Registry.

**The Idea here is `not` to talk down Apache Mynewt Config, as its simplicity of course has its own advantages, but to point out more clearly how our solution differs to it.**

| Feature | Mynewt Config | RIOT Registry |
|--|--|--|
| Pointer based API | ❌ | ✅ |
| Integer Path based API | ❌ | ✅ |
| String Path based API | ✅ | ✅ |
| Shared Configuration Schemas with instances | ❌ | ✅ |
| Nested configuration groups / parameters | ✅ | ✅ |
| Parameter types (string, int8, uint32, float, ...) | ❌ | ✅ |
| Internal parameter value format | String | Any (defined by schema) |
| Persistent configuration | ✅ | ✅ |
| Applying multiple changes at once | ✅ | ✅ |

## External Configuration Managers

### CLI

The only available Configuration Manager in this PR is a CLI.\
It can be tried in the example under `examples/registry`.

The CLI uses int paths separated by `/`.\
E.g.: `0/0/0/0`.

### CoAP API

The idea is to use the `registry_int` and/or `registry_path` module to provide a simple CoAP API.

#### `/get` endpoint

The get function could be mapped to the CoAP GET function.

#### `/set` endpoint

The set function could be mapped to the CoAP PUT function.

#### `/apply` endpoint

This is more tricky.
Possible solutions could be to provide an `/apply` suffix or prefix at the end or beginning of a path, to tell the registry, that this is an apply operation.

#### `/export` endpoint

The export function could be implemented using a CoAP GET function using `/export` as a suffix or prefix of the path.
CoAP would then return one response containing all the requested objects for example structured using CBOR.

### LwM2M

A LwM2M integration would require to write a mapping between each LwM2M Object to the respective RIOT Registry Configuration Schemas.

In this case every LwM2M set operation would immediately trigger a `registry_apply` as LwM2M does not provide an `apply` operation.
This is not a drawback, as LwM2M allows to set multiple values at the same time, thus covering the same use-case as the `registry_apply` function.

### Testing procedure

Testing commands unter tests/unittests:

```shell
make tests-registry term

make tests-registry_storage term

make tests-registry_storage_heap term

make tests-registry_storage_vfs term

make tests-registry_int_path term

make tests-registry_string_path term
```

### Issues/PRs references

See also #10622
See also #19557

waehlisch commented:
thanks for this very careful PR description, including the feature matrix comparison!

maribu reviewed (changes requested):
Sorry for not having got any feedback on this so far.

Based on the metrics only, I doubt this has a chance to get merged. The amount of lines of codes IMO is way too large to justify adding it.

Pro arguments of merging this
- lots of cool features

Contra arguments of merging this
- maintaining 11K loc is a huge effort
- 11K loc translate into huge growth of `.text`, making it difficult to actually use on MCUs with highly constrained memory
- the typical use cases likely can be addressed well with the EEPROM registry that is already upstream, the FlashDB package that is already upstream, and the file systems that are already upstream

I'm aware that this is an a****e move from me to NACK this without having looked at the code. I'm pretty sure that most if not all of the other maintainers feel similar and didn't start to look into the code based on the metrics. And I think it is better to have frustrating feedback than having this lingering here and being ignored.

LasseRosenow commented:
> Based on the metrics only, I doubt this has a chance to get merged. The amount of lines of codes IMO is way too large to justify adding it.

Thanks for your reply, conveying your concerns regarding the peer size of this PR.

I want to do some clarification regarding the actual code size of this PR.  -- In-Fact most of my code lies in `tests` and also the core module itself is only `1.668` lines long and also includes a bunch of ìfdef`s that would usually not make their way into the compiled code. These macros make the code quite a lot larger, but they were necessary to add checks regarding minimum maximum and allowed values.

We could discuss removing this feature. It should default to be turned off at the moment anyways since I would argue, that very constrained devices don't want to waste processing time on these checks.

Okay to give more insight here is a list of the different modules of my implementation and their sizes:

### Configuration Schemas (Code-Generated)

| Module | Size in lines of code  | Explanation | 
|---|---|---|
| Registry Core | 1,668 | Basic functionality (already works good enough for `LwM2M` mappings |
| Persistent Storage Extension | 406 | Infrastructure to allow to write to `Storages` |
| Int-Path Extension | 695 | Totally optional (Allows to `find` Schemas using an integer path |
| String-Path Extension | 943 | Totally optional (Allows to `find` Schemas using a string path |

### Configuration Schemas (Code-Generated)

Configuration Schemas are generated by a python script and usually only the schema that is used will be compiled.

| Module | Size in lines of code  | Explanation | 
|---|---|---|
| Sys-Namespace | 539 | Contains a basic RGB LED Configuration Schema |
| Tests-Namespace | 3,146 | Contains many large configuration Schemas for various testing scenarios | 

### Storage Implementations

| Module | Size in lines of code  | Explanation | 
|---|---|---|
| Heap-Storage | 119 | Example storage implementation using heap memory |
| VFS-Storage | 349 | Example VFS based storage implementation |

### Applications

| Module | Size in lines of code  | Explanation | 
|---|---|---|
| Example Application | 158 | Shows how to use the Registry API |
| CLI Application | 501 | Registers a CLI to the RIOT Shell as an interface to the Registry |

### Tests

| Module | Size in lines of code
|---|---|
| Core Test | 1,987 | 
| Persistent Storage Test: 226 |
| Int-Path Test  315 |
| String-Path Test  285 |
| Heap-Storage Test | 156 |
| Storage-VFS Test | 183 |

OlegHahm commented:
From my perspective the lines of code are much less relevant compared to the added ROM requirements. Well, yes, this is an optional feature, but it still doesn't do any good when people can't use it in the majority of the cases because ROM would overflow.

I only did a very quick test comparing the provided registry example application with a similar application without it: the difference (on nrf52) was ~60k vs ~22k of ROM - that's indeed huge. However, of course this comparison is not quite fair and very superficial. Do you have more elaborated data on how big the actual overhead compared to alternate solutions would be?

benpicco commented:
Hey, sorry for the reactions being so negative. The problem with the config system is that *everybody* wants to have a config system for their subsystem, we can expect that a config system will be part of most firmwares in the future, so it's overhead can not be *too* big.

I understand that this PR includes all possible and optional features, but to better reason about the solution, can you provide a minimal viable config system core that just includes the basic functionality (without any optional dependencies).

LasseRosenow commented:
> Hey, sorry for the reactions being so negative. The problem with the config system is that *everybody* wants to have a config system for their subsystem, we can expect that a config system will be part of most firmwares in the future, so it's overhead can not be *too* big.
> 
> I understand that this PR includes all possible and optional features, but to better reason about the solution, can you provide a minimal viable config system core that just includes the basic functionality (without any optional dependencies).
> 
> We should also understand that we might to reach the *perfect* solution on the fist try, but if the API is sane we can always improve the implementation later on. 

Yeah no worries that is already on my list for early January, sorry that I haven't really found time to provide the more detailed measurements yet, it became a bit more busy for me towards the end of the year and Christmas :)

I think there is a need for me to provide better explanations of what each extension is used for and what their costs are.

I was planning to make a small user story for a minimal registry that for example gives some basic ability to manage the on board LED.

Anyways thank you for your feedback, it is very much appreciated by me :)

benpicco commented:
Alternatively you could also see if #19557 covers your use case and if not, what's lacking - @fabian18 is getting paid to work on this and will likely maintain it for the years to come, so without having looked into either PR extensively, that would me more inclined towards his solution.
On the other hand, having less work on our side is always welcome too, so if you think your solution is better, go ahead with it.

waehlisch commented:
> Alternatively you could also see if #19557 covers your use case and if not, what's lacking

to not lose track of ongoing discussions elsewhere: https://forum.riot-os.org/t/runtime-configuration-breakout-session/4006

LasseRosenow commented:
So I have added a minimal example application under `examples/registry_core`, which continuously toggles the `BOARD LED` `on` and `off`.

Running `make cosy` on it shows the following results:

LasseRosenow commented:
I have updated my comment before, because I was compiling for `native` before and have now changed the `BOARD` to `same54-xpro`  for better comparison.

MichelRottleuthner reviewed (changes requested):
Wooha many files. Thanks for the contribution and providing additional data :+1:
First round of review below.

MichelRottleuthner commented on sys/registry/namespace/sys/definitions/0001_rgb_led.yaml:
```diff
@@ -0,0 +1,40 @@
+        description: Brightness of the white color of the rgb lamp.
+        type: uint8
+
+  - id: 6
+    name: string
+    description: Intensity of the blue color of the rgb lamp.
+    type: string
+    size: 20
```
Is that a leftover?

MichelRottleuthner commented on sys/registry/namespace/sys/definitions/0001_rgb_led.yaml:
```diff
@@ -0,0 +1,40 @@
+    type: uint8
+
+  - id: 3
+    name: brightnesses
```
I'd go with the singular form as you also did in the description

LasseRosenow replied:
I removed the brightnesses attogether because it is not really part of an RGB LED anyways.

MichelRottleuthner commented on sys/registry/namespace/sys/namespace_sys_board_led.c:
```diff
@@ -0,0 +1,83 @@
+    }
+}
+
+/* Schema */
```
Nit: could be considered as part of the schema but I was confused that it says schema here and the actual schema type is further down.

LasseRosenow replied:
It could not only be considered as part of the schema, it IS part of the schema.

But I will try to improve the comments to make it not confusing.

MichelRottleuthner reviewed (changes requested):
some more.

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+
+/**
+ * @brief Registers a new storage as a destination for saving configurations
+ * Only one storage can be registered as destination at a time. If a
```
The reason, why I decided to do this was that if I have multiple destinations, I need some way to specify what `Configuration Parameter` should be written to what storage.

This certainly is an area that can be further explored if people find it useful though.
But for now I choose this as `an easy way out for now` :sweat_smile:

LasseRosenow replied:
This limitation is lifted now.

The `registry_storage_instance_t` is now always passed to the `registry_storage_load` or `registry_storage_save` function.

This makes the API more flexible and allows the application to decide from/to what storage to read or write.

I also introduced a `registry_storage_set_instances` and `registry_storage_get_instances` function that allows external configuration managers such as a CLI see the available storage instances.
The user of the CLI can then for example print them out and then decide which one to use when calling the `load` or `save` argument.

LasseRosenow commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @return Length of string path on success, non-zero on failure.
+ */
+int registry_to_schema_string_path(const registry_schema_t *schema, char *path);
```
I agree with moving the `_to_`, but I would prefer not to drop the string.
This would make the naming scheme inconsistent and less precise from my view.

So maybe: `registry_schema_to_string_path`
Or maybe make it more clear: `registry_convert_schema_to_string_path` (even though this is more long)

But if we choose to clean up all those many convert functions and move to a single function with a unified type, we could also have something like `registry_convert_to_string_path` or `registry_to_string_path`.

LasseRosenow replied:
I have opted for:

`registry_node_to_string_path`
and
`registry_node_from_string_path`

LasseRosenow replied:
This is also implemented already.

LasseRosenow commented on sys/include/registry/namespace/sys/rgb_led.h:
```diff
@@ -0,0 +1,65 @@
+    REGISTRY_SYS_RGB_LED_BRIGHTNESSES,
+    REGISTRY_SYS_RGB_LED_BRIGHTNESSES_WHITE,
+    REGISTRY_SYS_RGB_LED_BRIGHTNESSES_YELLOW,
+} registry_sys_rgb_led_indices_t;
```
So I could showcase the `group` functionality.
As you can see the `brightness` group contains the `white` and `yellow` brightness parameters.

And yes you are correct. There should also be a plain rgb_led schema and the naming of this schema is indeed not ideal.
I should better rename it to `rgbwy_led`. In theory `led` itself is also not so correct and maybe this should be changed to `rgbwy_light`?

LasseRosenow replied:
I made the example a basic RGB LED for now

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int registry_save(void);
```
See my comment on your next comment at line 140 for more details regarding this.

LasseRosenow replied:
The storage_instance is now a parameter of the `storage_save` and `storage_load` functions as you suggested.

LasseRosenow commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @return Length of string path on success, non-zero on failure.
+ */
+int registry_to_namespace_string_path(const registry_namespace_t *namespace, char *path);
```
This is not ideal. I will fix it.

LasseRosenow replied:
Done

LasseRosenow commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int registry_from_namespace_string_path(const char *path, registry_namespace_t **namespace);
```
I will think of this further once we agreed on how to handle the whole situation with all those registry_from_to_etc functions.

But I agree having only `_to_` seems attractive to me as well.

At that time I opted for the `_to_` and `_from_` because I could not think of a generic name for the basic registry pointer types.

Example:
```c
registry_to_string_path();
registry_from_string_path();

// vs.

registry_to_string_path();
registry_string_path_to_registry();

// vs.

registry_registry_to_string_path();
registry_string_path_to_registry();

// vs.

registry_pointer_to_string_path();
registry_string_path_to_pointer();
```

I am happy with none of them :cry:

LasseRosenow replied:
This is totally outdated and does not exist anymore.
So it is "fixed"

LasseRosenow commented on sys/include/registry/util.h:
```diff
@@ -0,0 +1,65 @@
+
+#include "registry.h"
+
+void _debug_print_value(const registry_value_t *value);
```
I should put it in a #ifdef DEBUG

It is for testing if debug is enabled.

LasseRosenow replied:
Let me correct myself. This was devel code and is now removed.

LasseRosenow commented on sys/registry/Kconfig:
```diff
@@ -0,0 +1,42 @@
+    bool "Include description string in schemas"
+
+config REGISTRY_ENABLE_ALLOWED_VALUES_CHECK
+    bool "Check if new values are allowed"
```
Yes this configures if the registry should check for allowed values.

When we call `registry_set` we can pass what ever value we want. But some Schemas specify which values are actually allowed.

LasseRosenow replied:
This is the way I handle ENUMs for example

LasseRosenow replied:
This does not exist anymore. I decided that these kind of checks should be done on the side of the configuration manager and not on the constrained device.

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+    return -REGISTRY_ERROR_INSTANCE_NOT_FOUND;
+}
+
+static const registry_group_t *_internal_group_lookup(const registry_group_t *group,
```
I agree this is on my TODO list.

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+}
+
+registry_instance_int_path_t registry_to_instance_int_path(const registry_instance_t *instance)
+{
+    assert(instance != NULL);
+
+    return (registry_instance_int_path_t) {
+               .namespace_id = instance->schema->namespace->id,
+               .schema_id = instance->schema->id,
+               .instance_id = instance->id,
+    };
+}
+
+registry_group_int_path_t registry_to_group_int_path(const registry_instance_t *instance,
+                                                     const registry_group_t *group)
+{
+    assert(instance != NULL);
+    assert(group != NULL);
+
+    return (registry_group_int_path_t) {
+               .namespace_id = instance->schema->namespace->id,
+               .schema_id = instance->schema->id,
+               .instance_id = instance->id,
+               .group_id = group->id,
+    };
+}
+
+registry_parameter_int_path_t registry_to_parameter_int_path(const registry_instance_t *instance,
+                                                             const registry_parameter_t *parameter)
+{
+    assert(instance != NULL);
+    assert(parameter != NULL);
+
+    return (registry_parameter_int_path_t) {
+               .namespace_id = instance->schema->namespace->id,
+               .schema_id = instance->schema->id,
+               .instance_id = instance->id,
+               .parameter_id = parameter->id,
+    };
+}
```
Yes we hopefully can overcome this when we rethink all these functions soon.

LasseRosenow replied:
This got reworked now

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+    /* leaving namespace, schema or instance NULL is fine */
+    assert(group != NULL);
+    assert(parameter != NULL);
+
+    registry_namespace_t *found_namespace;
+    int res = _namespace_lookup(path->namespace_id, &found_namespace);
+    if (namespace != NULL) {
+        *namespace = found_namespace;
+    }
+
+    if (res == 0) {
+        registry_schema_t *found_schema;
+        res = _schema_lookup(found_namespace, path->schema_id, &found_schema);
+        if (schema != NULL) {
+            *schema = found_schema;
+        }
+
+        if (res == 0) {
+            registry_instance_t *found_instance;
+            res = _instance_lookup(found_schema, path->instance_id, &found_instance);
+            if (instance != NULL) {
+                *instance = found_instance;
+            }
+
+            if (res == 0) {
+                res = _parameter_lookup(found_schema, path->group_or_parameter_id, parameter);
+                if (res == 0) {
+                    *path_type = REGISTRY_INT_PATH_TYPE_PARAMETER;
+                }
+                else {
+                    res = _group_lookup(found_schema, path->group_or_parameter_id, group);
+                    if (res == 0) {
+                        *path_type = REGISTRY_INT_PATH_TYPE_GROUP;
+                    }
+                }
+            }
+        }
+    }
+
+    return res;
```
Yes this will be reworked.

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+                                  registry_namespace_t **namespace, registry_schema_t **schema)
+{
+    assert(path != NULL);
+    /* leaving namespace NULL is fine */
```
It is fine, because `namespace` is an output parameter.
If the user of the API is only interested in the `schema`, then the user can just put `NULL` to the namespace as a way of telling the API: I am not interested in the namespace, just give me the schema.

As with the other issues before. Once I changed this to have only one function to rule them all, we will automatically fix this as well. (Because there will only be one `registry_node_t` or similar name? output struct. instead of having `namespace` and `registry` separately.

LasseRosenow replied:
This is outdated and fixed. This function does not exist anymore

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+_REGISTRY_CREATE_PARAMETER_CONSTRAINT_TYPE_ALLOWED_FORBIDDEN_MIN_MAX(int64, int64_t)
+
+_REGISTRY_CREATE_PARAMETER_CONSTRAINT_TYPE_ALLOWED_FORBIDDEN_MIN_MAX(float32, float)
+_REGISTRY_CREATE_PARAMETER_CONSTRAINT_TYPE_ALLOWED_FORBIDDEN_MIN_MAX(float64, double)
```
I decided to remove this kind of checks altogether.
I think the huge amount of macro magic is not worth it and hard to maintain.

In general doing these kind of checks on constrained devices is controversial.
We should consider implementing this on the `Configuration Manager`s side.

So this is resolved as of my recent fixups.

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+                                 const void *context)
+{
+    (void)scope;
+    (void)context;
```
For the registry cli the `context` parameter really has no use, so I cannot find a good way to showcase it, but I will do something in the other example.

LasseRosenow replied:
Apart from that this should be fixed

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+        printf("whole instance");
+    }
+
+    printf("\n");
```
Fixed

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+#include "registry/namespace/tests.h"
+#include "registry/namespace/tests/nested.h"
+
+int rgb_led_instance_0_commit_cb(const registry_commit_cb_scope_t scope,
```
I decided to just rename it and added some explanation as to why both use the same function.

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+REGISTRY_ADD_STORAGE_SOURCE(vfs_instance);
+REGISTRY_SET_STORAGE_DESTINATION(vfs_instance);
+
+static registry_tests_nested_instance_t test_nested_instance_data = {
```
I added some comments to explain more in the whole example.

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+};
+
+static registry_instance_t test_nested_instance = {
+    .name = "instance-1",
```
I removed this instance, because it uses a schema that I want to move to the unit test folder.

LasseRosenow commented on examples/registry_cli/main.c:
```diff
@@ -0,0 +1,127 @@
+    registry_add_schema_instance(&registry_tests_nested, &test_nested_instance);
+
+    /* init storage */
+    #if IS_USED(MODULE_LITTLEFS2)
```
Done :)

LasseRosenow commented on examples/registry_core/Makefile:
```diff
@@ -0,0 +1,22 @@
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../..
+
+# required modules
```
Done

LasseRosenow commented on examples/registry_core/main.c:
```diff
@@ -0,0 +1,107 @@
+ * @{
+ *
+ * @file
+ * @brief       RIOT Registry example application
```
Done

LasseRosenow commented on examples/registry_core/main.c:
```diff
@@ -0,0 +1,107 @@
+    (void)scope;
+    (void)context;
+
+    if (group_or_parameter_id != NULL) {
+        if (*group_or_parameter_id == REGISTRY_SYS_BOARD_LED_ENABLED) {
+            /* The Driver owns the board_led data instance, so we can just get our value from there */
+            bool led_state = board_led_instance_0_data.enabled;
+
+            /* Turn the LED on or off depending on the led_state */
+            if (led_state == true) {
+                /* This is the commit_cb function of instance 0, so we toggle LED 0 as well */
+                LED_ON(0);
+            } else {
+                LED_OFF(0);
+            }
+        }
+    }
+    else {
+        /* The whole instance got committed in one go, so apply all parameters (BOARD_LED has only one anyways)*/
+
+        bool led_state = board_led_instance_0_data.enabled;
+        if (led_state == true) {
+            LED_ON(0);
+        } else {
+            LED_OFF(0);
+        }
+    }
```
Done

LasseRosenow commented on examples/registry_core/main.c:
```diff
@@ -0,0 +1,107 @@
+        /* Set new registry value */
+        registry_set(&board_led_instance, &registry_sys_board_led_enabled, &board_led_enabled, sizeof(board_led_enabled));
+
+        /* Apply the registry value to change the LED state (calls the commit_cb function implemented by the BOARD for example)*/
```
Done

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+} registry_commit_cb_scope_t;
+
+/**
+ * @brief Callback must be implemented by modules / drivers that integrate a configuration schema.
```
Done

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+
+
+#if IS_ACTIVE(CONFIG_REGISTRY_ENABLE_ALLOWED_VALUES_CHECK) || IS_ACTIVE(DOXYGEN)
+    #define _REGISTRY_CREATE_PARAMETER_CONSTRAINT_TYPE_PART_ALLOWED_VALUES(_type) \
```
I removed the allowed values feature entirely

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+     * @param[in] parameter_id ID of the parameter that contains the value.
+     * @param[in] instance Pointer to the instance of the schema, that contains the parameter.
+     * @param[in] val Pointer to buffer containing the new value.
+     * @param[in] val_len Pointer to length of the buffer to store the current value.
```
Fixed

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+} registry_export_cb_data_t;
+
+typedef const enum {
+    REGISTRY_EXPORT_NAMESPACE,
```
This enum does not exist anymore after the latest major refactor, so this is implicitly fixed

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+} registry_value_t;
+
+typedef const enum {
+    REGISTRY_COMMIT_INSTANCE,
```
Done

LasseRosenow commented on sys/include/registry/error.h:
```diff
@@ -0,0 +1,45 @@
+#endif
+
+/**
+ * @brief Error codes start at 1_000 to prevent clashing with stdlib error codes.
```
Anyways I changed the error codes to start from 1, so this is not an issue anymore

LasseRosenow commented on sys/include/registry/error.h:
```diff
@@ -0,0 +1,45 @@
+ * @brief Error codes start at 1_000 to prevent clashing with stdlib error codes.
+ */
+typedef const enum {
+    REGISTRY_ERROR_NONE = 1000,
```
Done, but migrating the registry to use its custom error codes is not yet completed

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+
+typedef int (*registry_export_cb_t)(const registry_export_cb_data_t *data,
+                                    const registry_export_cb_data_type_t data_type,
+                                    const void *context);
```
Done

LasseRosenow commented on sys/include/registry.h:
```diff
@@ -0,0 +1,532 @@
+ * @brief Instance of a schema containing its data.
+ */
+struct _registry_instance_t {
+    clist_node_t node;                  /**< Linked list node. */
```
I changed the comment accordingly

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+/**
+ * @defgroup    sys_registry_int_path RIOT Registry integer path
+ * @ingroup     sys
+ * @brief       RIOT Registry integer path module providing functions to convert between registry objects and their integer paths
```
Done

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+/**
+ * @brief Maximum length of a configuration path as a string.
+ *
+ * A int path ID is an uint32_t and uint32_t MAX has 10 digits.
```
Obsolete (Fixed)

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+/**
+ * @brief Maximum length of a configuration path as a string.
+ *
+ * A int path ID is an uint32_t and uint32_t MAX has 10 digits.
```
Done

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+    registry_namespace_id_t namespace_id;                   /**< The ID of the namespace. */
+    registry_schema_id_t schema_id;                         /**< The ID of the schema. */
+    registry_instance_id_t instance_id;                     /**< The ID of the instance. */
+    registry_group_or_parameter_id_t group_or_parameter_id; /**< The ID of the group or parameter. */
```
Done

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+    registry_schema_id_t schema_id;                         /**< The ID of the schema. */
+    registry_instance_id_t instance_id;                     /**< The ID of the instance. */
+    registry_group_or_parameter_id_t group_or_parameter_id; /**< The ID of the group or parameter. */
+} registry_group_or_parameter_int_path_t;
```
I have now migrated the int_path module to only have 2 functions.
One to convert to an `int_path_t` to `registry_node_t` and one to convert a `registry_node_t` to an `int_path_t`.
The `int_path_t` as of now is a union over the different path types.

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+ * @return The integer path representation of the parameter.
+ */
+registry_parameter_int_path_t registry_to_parameter_int_path(const registry_instance_t *instance,
+                                                             const registry_parameter_t *parameter);
```
This is now implemented as requested and stated in an earlier comment:

I have now migrated the int_path module to only have 2 functions.
One to convert to an `int_path_t` to `registry_node_t` and one to convert a `registry_node_t` to an `int_path_t`.
The `int_path_t` as of now is a union over the different path types.

LasseRosenow commented on sys/include/registry/int_path.h:
```diff
@@ -0,0 +1,266 @@
+                                              registry_schema_t **schema,
+                                              registry_instance_t **instance,
+                                              registry_group_t **group,
+                                              registry_parameter_t **parameter);
```
This is now implemented with a unified `registry_int_path_t` and `registry_node_t` type.

LasseRosenow commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,183 @@
+                                                 registry_schema_t **schema,
+                                                 registry_instance_t **instance,
+                                                 registry_group_t **group,
+                                                 registry_parameter_t **parameter);
```
There is a difference of course, but anyways this function does not exist anymore.

LasseRosenow commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,183 @@
+#include "registry.h"
+
+/**
+ * @brief Converts a registry namespace object to its string path representation.
```
Done

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+                return 0;
+            }
+        }
+    }
```
This is fixed now

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+        node = node->next;
+
+        /* check if index equals instance_id */
+        if (container_of(node, registry_instance_t, node)->id == instance_id) {
+            *instance = container_of(node, registry_instance_t, node);
```
Done

LasseRosenow commented on sys/registry/int_path.c:
```diff
@@ -0,0 +1,426 @@
+/**
+ * @defgroup    sys_registry_int_path RIOT Registry Int Path
+ * @ingroup     sys
+ * @brief       RIOT Registry Int Path module providing a API to access configuration parameter via an integer path
```
Done

LasseRosenow commented on sys/registry/Makefile.dep:
```diff
@@ -0,0 +1,14 @@
+endif
+
+USEMODULE += fmt
+USEMODULE += base64
```
I removed this and put util into its own module

LasseRosenow commented on sys/registry/init.c:
```diff
@@ -0,0 +1,51 @@
+#include <assert.h>
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
```
I removed them

LasseRosenow commented on sys/registry/Makefile.dep:
```diff
@@ -0,0 +1,14 @@
+  USEMODULE += registry
+endif
+
+USEMODULE += fmt
```
This is fixed

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int registry_load(void);
```
Done

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @return 0 on success, non-zero on failure.
+     */
+    int (*save_end)(const registry_storage_instance_t *storage);
```
Done

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+     *
+     * @return 0 on success, non-zero on failure.
+     */
+    int (*save_start)(const registry_storage_instance_t *storage);
```
Done

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+ */
+typedef struct {
+    registry_storage_t *storage;        /**< interface of the storage. */
+    void *data;                         /**< Struct containing all config data for the storage. */
```
Doc was added

LasseRosenow commented on sys/include/registry/storage.h:
```diff
@@ -0,0 +1,164 @@
+ * @brief Prototype of a callback function for the load action of a storage interface.
+ */
+typedef int (*load_cb_t)(const registry_instance_t *instance, const registry_parameter_t *parameter,
+                         const void *buf, const size_t buf_len);
```
Done

LasseRosenow commented on sys/include/registry/namespace/tests/nested.h:
```diff
@@ -0,0 +1,61 @@
+
+typedef struct {
+    clist_node_t node;
+
```
Fixed

LasseRosenow commented on sys/include/registry/namespace/tests/nested.h:
```diff
@@ -0,0 +1,61 @@
+
+typedef struct {
+    uint8_t value;
+} registry_tests_nested_instance_opaque_t;
```
That was leftover devel code. It is gone now

LasseRosenow commented on sys/include/registry/namespace/tests/nested.h:
```diff
@@ -0,0 +1,61 @@
+/**
+ * @defgroup    sys_registry_namespace_tests_nested RIOT Registry Schema: Nested
+ * @ingroup     sys
+ * @brief       RIOT Registry Nested Schema representing different nesting level of a configuration schema
```
Fixed

LasseRosenow commented on sys/include/registry/namespace/tests/full.h:
```diff
@@ -0,0 +1,98 @@
+/**
+ * @defgroup    sys_registry_namespace_tests_full RIOT Registry Schema: Full
+ * @ingroup     sys
+ * @brief       RIOT Registry Full Schema representing all possible data types of the riot registry
```
Fixed

LasseRosenow commented on sys/include/registry/namespace/tests.h:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @defgroup    sys_registry_namespace_tests RIOT Registry Tests Namespace
+ * @ingroup     tests
+ * @brief       RIOT Registry Namespace Tests module providing common tests configuration schemas for the RIOT Registry sys module
```
Fixed

LasseRosenow commented on sys/include/registry/namespace/sys.h:
```diff
@@ -0,0 +1,41 @@
+
+extern registry_namespace_t registry_sys;
+
+typedef enum {
```
Done (This file is code generated, so I only give some basic explanation

maribu reviewed (commented):
Thanks for updating the PR and providing the metrics.

I see that the LoC have now gone down to 7K. This is still rather large for reviewing. Would it be possible if you split out a PR that does not provide any storage backend yet? E.g. just the base features + test + an example that uses the shell interface, no persistence yet?

I added some nitpicks inline.

AnnsAnns commented:
Thank you for the rebase, lets get this rolling again :steam_locomotive:🚃🚃🚃

AnnsAnns commented on sys/include/registry/error.h:
```diff
@@ -0,0 +1,49 @@
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
```suggestion
/*
 * SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 * SPDX-FileCopyrightText: 2025 HAW Hamburg
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

AnnsAnns commented on sys/include/registry/namespace/sys/board_led.h:
```diff
@@ -0,0 +1,51 @@
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
```suggestion
/*
 * SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 * SPDX-FileCopyrightText: 2025 HAW Hamburg
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

AnnsAnns commented on sys/include/registry/string_path.h:
```diff
@@ -0,0 +1,61 @@
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
```suggestion
/*
 * SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 * SPDX-FileCopyrightText: 2025 HAW Hamburg
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

AnnsAnns commented on sys/registry/Kconfig:
```diff
@@ -0,0 +1,30 @@
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
```
```suggestion
 # SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 # SPDX-FileCopyrightText: 2025 HAW Hamburg
 # SPDX-License-Identifier: LGPL-2.1-only
```

AnnsAnns commented on sys/registry/find.c:
```diff
@@ -0,0 +1,287 @@
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
```suggestion
/*
 * SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 * SPDX-FileCopyrightText: 2025 HAW Hamburg
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

AnnsAnns commented on sys/registry/namespace/sys/namespace_sys.c:
```diff
@@ -0,0 +1,53 @@
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
```
```suggestion
/*
 * SPDX-FileCopyrightText: 2025 Lasse Rosenow <lasse.rosenow@haw-hamburg.de>
 * SPDX-FileCopyrightText: 2025 HAW Hamburg
 * SPDX-License-Identifier: LGPL-2.1-only
 */
```

AnnsAnns commented on sys/include/registry.h:
```diff
@@ -0,0 +1,403 @@
+ */
+typedef struct {
+    registry_type_t type; /**< The type of the configuration parameter value. */
+    const void *buf;      /**< Pointer to the buffer containing the value of the configuration parameter. */
```
Long line

AnnsAnns commented on sys/include/registry.h:
```diff
@@ -0,0 +1,403 @@
+            const registry_instance_t *instance;   /**< Pointer to a schema instance */
+            const registry_parameter_t *parameter; /**< Pointer to a configuration parameter */
+        } parameter;                               /**< A configuration parameter depends on an instance */
+    } value;                                       /**< The location inside of the configuration tree */
```
Generally I'm fine with these type of "long lines" where it's long because it tries to conform to a comment formatting but I'm not sure what the general consensus on this is (Maybe @crasbe for a short confirmation whether or not this counts as a long line)

AnnsAnns commented on sys/include/registry.h:
```diff
@@ -0,0 +1,403 @@
+
+/**
+ * @brief Instance of a schema containing its configuration parameters values.
+ * The instance of a configuration schema is created by the modules that need to expose runtime configurations.
```
```suggestion
 * The instance of a configuration schema is created by the modules
 * that need to expose runtime configurations.
```

AnnsAnns commented on sys/registry/storage/storage.c:
```diff
@@ -0,0 +1,107 @@
+{
+    const registry_storage_instance_t *storage_instance = context;
+
+    /* the registry also exports just the namespace or just a schema, but the storage is only interested in configuration parameter values */
```
Really long line

AnnsAnns reviewed (changes requested):
Note, I mostly only tried to get a basic overview (talked about below).
These are mostly nitpicks, I can happily say that most of the documentation is quite easy to understand and the amount of documentation is definitely a positive :smile: 

Also happy to see so many different unit tests and in-function assertions. It does definitely give me confidence in the code maintainability.

That being said, if (depending on discussions) we want to move ahead with merging this, I do have to agree that this PR should most likely be split into a few separate PRs.
While the code is well documented and strongly adheres to the coding conventions, it's simply *a lot* of LoC.

Just looking at the unit tests alone might feel frightening and looking at everything in one gigantic PR makes it feel really overwhelming. Nonetheless thank you for your work on rebasing this and applying newer/modern RIOT coding convention guidelines.

LasseRosenow commented:
I close this, as the plan is to continue this in smaller reviewable pull requests.

The first one is: #22092

After that one got merged, I will rebase this PR to keep a single place with everything working and then create more split-out PRs based on this one.

LasseRosenow committed:
sys/registry: add runtime config registry

LasseRosenow commented:
> > After that one got merged, I will rebase this PR to keep a single place with everything working and then create more split-out PRs based on this one.
> 
> In that case I'd rather leave it open as draft PR :)

@crasbe this PR was marked as "draft" on purpose, because this is still the place that contains all the changes which step by step will be branched out in single PRs and then rebased back here, when they are merged in main.

LasseRosenow committed:
Update
